classes/external/list_potential_courses.php

```diff
@@ -0,0 +1,89 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace availability_othercompleted\external;
+
+use context_course;
+
+/**
+ * Gets potential courses for form autocomplete.
+ * @package   availability_othercompleted
+ * @author    Matthew Hilton (matthewhilton@catalyst-au.net)
+ * @copyright Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class list_potential_courses extends \external_api {
+
+    /**
+     * Specifies parameters
+     * @return mixed external function parameters
+     */
+    public static function get_parameters() {
+        return new \external_function_parameters([
+            'query' => new \external_value(PARAM_TEXT, 'Query string to filter results', VALUE_DEFAULT, ''),
+            ]);
+    }
+
+    /**
+     * Lists potential courses from the site.
+     * @param string $query query string to query courses fullname
+     * @return array courses selected by the query string
+     */
+    public static function get($query = '') {
+        global $DB;
+
+        if ($query == '') {
+            return [];
+        }
+
+        $courses = $DB->get_records_select(
+            'course',
+            $DB->sql_like('fullname', ':name', false, false),
+            ['name' => '%'.$query.'%'],
+            '',
+            'id, fullname, enablecompletion'
+        );
+
+        $availablecourses = array_filter($courses, function($course) {
+            global $SITE;
+
+            // Filter these by courses they can manage.
+            $context = context_course::instance($course->id);
+            $canmanage = has_capability('moodle/course:update', $context);
+
+            // Ignore dashboard/site course.
+            $notsitecourse = $course->id != $SITE->id;
+
+            // Only those that have completion enabled.
+            $hascompleteionenabled = $course->enablecompletion == 1;
```

Comment:
This one feels like it could be bundled into the SQL, to reduce the loop count

Reply:
Good call - done!